File: app/src/utils/sparql.ts
import ParsingClient from 'sparql-http-client/ParsingClient';
import { NamedNode } from 'rdf-js';

const makeClient = () => {
  const endpointUrl = 'http://localhost:7200/repositories/kgrc4si';
  return new ParsingClient({
    endpointUrl: `${endpointUrl}?infer=false`,
  });
};

export const PREFIXES = {
  ex: 'http://kgrc4si.home.kg/virtualhome2kg/instance/',
  mssn: 'http://mssn.sigappfr.org/mssn/',
  rdf: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
  vh2kg: 'http://kgrc4si.home.kg/virtualhome2kg/ontology/',
};

export type ActivityQueryType = {
  activity: NamedNode;
};
export const fetchActivity: () => Promise<ActivityQueryType[]> = async () => {
  const query = `
PREFIX vh2kg: <http://kgrc4si.home.kg/virtualhome2kg/ontology/>
select DISTINCT ?activity where {
    ?activity vh2kg:hasVideo ?camera .
    filter (regex(str(?camera), "camera0"))
} order by asc(?activity)
  `;
  const result = (await makeClient().query.select(
    query
  )) as ActivityQueryType[];
  return result;
};

export type CameraQueryType = {
  camera: NamedNode;
};
export const fetchCamera: (
  activity: string,
  scene: string
) => Promise<CameraQueryType[]> = async (activity, scene) => {
  if (!activity || !scene) return Promise.resolve([]);

  const query = `
PREFIX vh2kg: <http://kgrc4si.home.kg/virtualhome2kg/ontology/>
PREFIX ex: <http://kgrc4si.home.kg/virtualhome2kg/instance/>
select DISTINCT ?camera where {
    ex:${activity}_${scene} vh2kg:hasVideo ?camera .
} order by asc(?camera)
  `;
  const result = (await makeClient().query.select(query)) as CameraQueryType[];
  return result;
};

export type EventQueryType = {
  event: NamedNode;
  action: NamedNode;
  startFrame: NamedNode;
};
export const fetchEvent: (
  activity: string,
  scene: string,
  camera: string
) => Promise<EventQueryType[]> = async (activity, scene, camera) => {
  if (!activity || !scene || !camera) return Promise.resolve([]);

  const query = `
PREFIX vh2kg: <http://kgrc4si.home.kg/virtualhome2kg/ontology/>
PREFIX ex: <http://kgrc4si.home.kg/virtualhome2kg/instance/>
PREFIX mssn: <http://mssn.sigappfr.org/mssn/>
select DISTINCT ?event ?startFrame ?action where {
    ex:${activity}_${scene}_${camera} mssn:hasMediaSegment ?segment .
    ?segment vh2kg:isVideoSegmentOf ?event ;
            vh2kg:hasStartFrame ?startFrame .
    ?event vh2kg:action ?action .
}order by asc(?startFrame)
  `;
  const result = (await makeClient().query.select(query)) as EventQueryType[];
  return result;
};

export type ObjectQueryType = {
  object: NamedNode;
};
export const fetchObject: (
  activity: string,
  scene: string,
  camera: string,
  event: string
) => Promise<ObjectQueryType[]> = async (activity, scene, camera, event) => {
  if (!activity || !scene || !camera) return Promise.resolve([]);

  let query = '';

  if (event) {
    query = `
PREFIX ex: <http://kgrc4si.home.kg/virtualhome2kg/instance/>
PREFIX mssn: <http://mssn.sigappfr.org/mssn/>
PREFIX vh2kg: <http://kgrc4si.home.kg/virtualhome2kg/ontology/>
select DISTINCT ?object where {
    ex:${activity}_${scene}_${camera} mssn:hasMediaSegment ?segment .
    ?segment vh2kg:isVideoSegmentOf ex:${event}_${activity}_${scene} ;
             mssn:hasMediaDescriptor ?descriptorFrame .
    ?descriptorFrame mssn:hasMediaDescriptor ?descriptor .
    ?descriptor vh2kg:is2DbboxOf ?object .
}order by asc(?object)
    `;
  } else {
    query = `
PREFIX ex: <http://kgrc4si.home.kg/virtualhome2kg/instance/>
PREFIX mssn: <http://mssn.sigappfr.org/mssn/>
PREFIX vh2kg: <http://kgrc4si.home.kg/virtualhome2kg/ontology/>
select DISTINCT ?object where {
    ex:${activity}_${scene}_${camera} mssn:hasMediaSegment ?segment .
    ?segment mssn:hasMediaDescriptor ?descriptorFrame .
    ?descriptorFrame mssn:hasMediaDescriptor ?descriptor .
    ?descriptor vh2kg:is2DbboxOf ?object .
}order by asc(?object)
  `;
  }
  const result = (await makeClient().query.select(query)) as ObjectQueryType[];
  return result;
};

export type StartFrameOfObjectQueryType = {
  startFrame: NamedNode;
};
export const fetchStartFrameOfObject: (
  activity: string,
  scene: string,
  camera: string,
  event: string,
  object: string
) => Promise<StartFrameOfObjectQueryType[]> = async (
  activity,
  scene,
  camera,
  event,
  object
) => {
  if (!activity || !scene || !camera || !object) return Promise.resolve([]);

  let query = '';

  if (event) {
    query = `
PREFIX ex: <http://kgrc4si.home.kg/virtualhome2kg/instance/>
PREFIX mssn: <http://mssn.sigappfr.org/mssn/>
PREFIX vh2kg: <http://kgrc4si.home.kg/virtualhome2kg/ontology/>
select DISTINCT ?startFrame where {
    ex:${activity}_${scene}_${camera} mssn:hasMediaSegment ?segment .
    ?segment vh2kg:isVideoSegmentOf ex:${event}_${activity}_${scene} ;
             mssn:hasMediaDescriptor ?startFrame .
    ?startFrame mssn:hasMediaDescriptor ?imageFrame .
    ?imageFrame vh2kg:is2DbboxOf ex:${object}_${scene} .
}order by asc(?imageFrame) limit 1
    `;
  } else {
    query = `
PREFIX ex: <http://kgrc4si.home.kg/virtualhome2kg/instance/>
PREFIX mssn: <http://mssn.sigappfr.org/mssn/>
PREFIX vh2kg: <http://kgrc4si.home.kg/virtualhome2kg/ontology/>
select DISTINCT ?startFrame where {
    ex:${activity}_${scene}_${camera} mssn:hasMediaSegment ?segment .
    ?segment mssn:hasMediaDescriptor ?startFrame .
    ?startFrame mssn:hasMediaDescriptor ?imageFrame .
    ?imageFrame vh2kg:is2DbboxOf ex:${object}_${scene} .
}order by asc(?imageFrame) limit 1
  `;
  }
  const result = (await makeClient().query.select(
    query
  )) as StartFrameOfObjectQueryType[];
  return result;
};

export type VideoQueryType = {
  frameRate: NamedNode;
  video: NamedNode;
};
export const fetchVideo: (
  activity: string,
  scene: string,
  camera: string
) => Promise<VideoQueryType[]> = async (activity, scene, camera) => {
  const query = `
PREFIX ex: <http://kgrc4si.home.kg/virtualhome2kg/instance/>
PREFIX vh2kg: <http://kgrc4si.home.kg/virtualhome2kg/ontology/>
select ?frameRate ?video where {
    ex:${activity}_${scene}_${camera} vh2kg:frameRate ?frameRate ;
                                      vh2kg:video ?video .
}
  `;
  const result = (await makeClient().query.select(query)) as VideoQueryType[];
  return result;
};

export type ImageQueryType = {
  image: NamedNode;
  splitWidth: NamedNode;
  resolution: NamedNode;
};
export const fetchImage: (
  activity: string,
  scene: string,
  camera: string,
  frame: number
) => Promise<ImageQueryType[]> = async (activity, scene, camera, frame) => {
  const query = `
PREFIX ex: <http://kgrc4si.home.kg/virtualhome2kg/instance/>
PREFIX vh2kg: <http://kgrc4si.home.kg/virtualhome2kg/ontology/>
PREFIX mssn: <http://mssn.sigappfr.org/mssn/>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
select ?splitWidth ?image ?resolution where {
    ex:${activity}_${scene}_${camera} mssn:hasMediaSegment ?segment ;
                                      vh2kg:hasResolution ?resolution .
    ?segment mssn:hasMediaDescriptor ?descriptor .
    ?descriptor vh2kg:frameNumber ${frame} ;
                vh2kg:splitWidth ?splitWidth ;
                vh2kg:image ?splitImage .
    ?splitImage vh2kg:splitImageID ?id ;
            rdf:value ?image .
}order by asc(?id)
  `;
  const result = (await makeClient().query.select(query)) as ImageQueryType[];
  return result;
};

File: cli/sparql.py
PREFIX_EX = "http://kgrc4si.home.kg/virtualhome2kg/instance/"
PREFIX_VH2KG = "http://kgrc4si.home.kg/virtualhome2kg/ontology/"
ENDPOINT = "http://localhost:7200/repositories/kgrc4si"


def get_all_frames(activity, scene, camera):
    print("Searching for all frames...")
    from SPARQLWrapper import SPARQLWrapper, JSON
    frame_list = {}

    query = """
PREFIX ex: <http://kgrc4si.home.kg/virtualhome2kg/instance/>
PREFIX vh2kg: <http://kgrc4si.home.kg/virtualhome2kg/ontology/>
PREFIX mssn: <http://mssn.sigappfr.org/mssn/>
select DISTINCT ?segment ?start_frame ?end_frame where {
    ex:""" + activity + "_" + scene + "_" + camera + """ mssn:hasMediaSegment ?segment.
    ?segment vh2kg:hasStartFrame ?start_frame ;
             vh2kg:hasEndFrame ?end_frame .
}
    """
    sparql = SPARQLWrapper(ENDPOINT)
    sparql.setQuery(query)
    sparql.setReturnFormat(JSON)
    results = sparql.query().convert()

    bindings = results["results"]["bindings"]
    for result in bindings:
        segment = result["segment"]["value"].replace(PREFIX_EX, "")
        frame_list[segment] = {'start_frame': int(result["start_frame"]["value"]), 'end_frame': int(result["end_frame"]["value"])}

    return frame_list


def get_frames_from_action(activity, scene, camera, action):
    print("Searching for frames from event...")
    from SPARQLWrapper import SPARQLWrapper, JSON
    frame_list = {}

    query = """
PREFIX ex: <http://kgrc4si.home.kg/virtualhome2kg/instance/>
PREFIX vh2kg: <http://kgrc4si.home.kg/virtualhome2kg/ontology/>
PREFIX mssn: <http://mssn.sigappfr.org/mssn/>
select DISTINCT ?segment ?start_frame ?end_frame where {
    ex:""" + activity + "_" + scene + "_" + camera + """ mssn:hasMediaSegment ?segment.
    ?segment vh2kg:isVideoSegmentOf ?event ;
             vh2kg:hasStartFrame ?start_frame ;
             vh2kg:hasEndFrame ?end_frame .
    ?event vh2kg:action ?action .
    filter (regex(str(?action), '""" + action + """'))
}
    """
    sparql = SPARQLWrapper(ENDPOINT)
    sparql.setQuery(query)
    sparql.setReturnFormat(JSON)
    results = sparql.query().convert()

    bindings = results["results"]["bindings"]
    for result in bindings:
        segment = result["segment"]["value"].replace(PREFIX_EX, "")
        frame_list[segment] = {'start_frame': int(result["start_frame"]["value"]), 'end_frame': int(result["end_frame"]["value"])}

    return frame_list


def get_frames_from_object(activity, scene, camera, action, object):
    print("Searching for frames from object...")
    from SPARQLWrapper import SPARQLWrapper, JSON
    frame_list = {}

    query = """
PREFIX ex: <http://kgrc4si.home.kg/virtualhome2kg/instance/>
PREFIX vh2kg: <http://kgrc4si.home.kg/virtualhome2kg/ontology/>
PREFIX mssn: <http://mssn.sigappfr.org/mssn/>
select DISTINCT ?segment ?frame_number where {
    ex:""" + activity + "_" + scene + "_" + camera + """ mssn:hasMediaSegment ?segment.
    ?segment mssn:hasMediaDescriptor ?descriptor .
    ?descriptor mssn:hasMediaDescriptor ?descriptor_object ;
                     vh2kg:frameNumber ?frame_number .
    ?descriptor_object vh2kg:is2DbboxOf ex:""" + object + "_" + scene + """ .
    """
    if action is not None:
        query += """
    ?event vh2kg:action ?action .
    filter (regex(str(?action), '""" + action + """'))
        """
    query += "} order by asc(?frame_number)"

    sparql = SPARQLWrapper(ENDPOINT)
    sparql.setQuery(query)
    sparql.setReturnFormat(JSON)
    results = sparql.query().convert()

    bindings = results["results"]["bindings"]
    frames = {}

    for result in bindings:
        segment = result["segment"]["value"].replace(PREFIX_EX, "")
        frame_number = int(result["frame_number"]["value"])
        if segment not in frames:
            frames[segment] = [frame_number]
        else:
            frames[segment].append(frame_number)

    for segment in frames:
        frame_list[segment] = {'start_frame': min(frames[segment]), 'end_frame': max(frames[segment])}

    return frame_list


def get_video(activity, scene, camera):
    print("Getting video...")
    from SPARQLWrapper import SPARQLWrapper, JSON

    query = """
PREFIX ex: <http://kgrc4si.home.kg/virtualhome2kg/instance/>
PREFIX vh2kg: <http://kgrc4si.home.kg/virtualhome2kg/ontology/>
select DISTINCT ?frame_rate ?video where {
    ex:""" + activity + "_" + scene + "_" + camera + """ vh2kg:frameRate ?frame_rate ;
                                                         vh2kg:video ?video .
}
    """
    sparql = SPARQLWrapper(ENDPOINT)
    sparql.setQuery(query)
    sparql.setReturnFormat(JSON)
    results = sparql.query().convert()

    return results


def get_images(segment, start_frame, end_frame):
    print("Getting images...")
    import base64
    import cv2
    import numpy
    from SPARQLWrapper import SPARQLWrapper, JSON
    image_dict = {}

    query = """
PREFIX ex: <http://kgrc4si.home.kg/virtualhome2kg/instance/>
PREFIX vh2kg: <http://kgrc4si.home.kg/virtualhome2kg/ontology/>
PREFIX mssn: <http://mssn.sigappfr.org/mssn/>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
select DISTINCT ?descriptor ?frame_number ?split_width ?image_id ?image where {
    ex:""" + segment + """ mssn:hasMediaDescriptor ?descriptor .
    ?descriptor vh2kg:frameNumber ?frame_number ;
                vh2kg:splitWidth ?split_width ;
                vh2kg:image ?split_image .
    ?split_image vh2kg:splitImageID ?image_id ;
                rdf:value ?image .
"""
    if start_frame is not None:
        query += "filter (?frame_number >= " + str(start_frame) + ")"
    if end_frame is not None:
        query += "filter (?frame_number <= " + str(end_frame) + ")"

    query += "} order by asc(?frame_number) asc(?image_id)"

    sparql = SPARQLWrapper(ENDPOINT)
    sparql.setQuery(query)
    sparql.setReturnFormat(JSON)
    results = sparql.query().convert()

    bindings = results["results"]["bindings"]
    for result in bindings:
        frame_number = int(result["frame_number"]["value"])
        if frame_number < start_frame or end_frame < frame_number:
            continue

        descriptor = result["descriptor"]["value"].replace(PREFIX_EX, "")
        split_width = int(result["split_width"]["value"])
        image_id = result["image_id"]["value"]
        base64_data = result["image"]["value"]
        if descriptor not in image_dict:
            image_dict[descriptor] = {'split_width': split_width, 'frame_number': frame_number, 'images': {}}
        img_raw = numpy.frombuffer(base64.b64decode(base64_data), numpy.uint8)
        image = cv2.imdecode(img_raw, cv2.IMREAD_UNCHANGED)
        image_dict[descriptor]['images'][image_id] = image

    for descriptor in image_dict:
        image_dict[descriptor]['images'] = dict(sorted(image_dict[descriptor]['images'].items(), key=lambda item: int(item[0])))

    return image_dict


def get_annotation_2d_bbox(scene, frame_list):
    print("Getting annotation 2D bbox...")
    from SPARQLWrapper import SPARQLWrapper, JSON
    annotation_list = []

    for segment in frame_list:
        if segment == 'all':
            continue

        start_frame = frame_list[segment]['start_frame']
        end_frame = frame_list[segment]['end_frame']
        query = """
PREFIX ex: <http://kgrc4si.home.kg/virtualhome2kg/instance/>
PREFIX vh2kg: <http://kgrc4si.home.kg/virtualhome2kg/ontology/>
PREFIX mssn: <http://mssn.sigappfr.org/mssn/>
select DISTINCT ?frame_number ?object ?2dbbox  where {
    ex:""" + segment + """ mssn:hasMediaDescriptor ?descriptor .
    ?descriptor mssn:hasMediaDescriptor ?descriptor_object ;
                     vh2kg:frameNumber ?frame_number .
    ?descriptor_object vh2kg:bbox-2d-value ?2dbbox ;
                       vh2kg:is2DbboxOf ?object .
"""
        if start_frame is not None:
            query += "filter (?frame_number >= " + str(start_frame) + ")"
        if end_frame is not None:
            query += "filter (?frame_number <= " + str(end_frame) + ")"

        query += "} order by asc(?frame_number)"

        sparql = SPARQLWrapper(ENDPOINT)
        sparql.setQuery(query)
        sparql.setReturnFormat(JSON)
        results = sparql.query().convert()

        bindings = results["results"]["bindings"]
        for result in bindings:
            frame_number = result["frame_number"]["value"]
            object = result["object"]["value"].replace(PREFIX_EX, "").replace("_" + scene, "")
            bbox = result["2dbbox"]["value"]
            annotation_list.append({'frame_number': frame_number, 'object': object, '2dbbox': bbox})

    return annotation_list


def get_annotation_action(scene, frame_list):
    print("Getting annotation action...")
    from SPARQLWrapper import SPARQLWrapper, JSON
    annotation_list = []

    for segment in frame_list:
        if segment == 'all':
            continue

        start_frame = frame_list[segment]['start_frame']
        end_frame = frame_list[segment]['end_frame']
        query = """
PREFIX ex: <http://kgrc4si.home.kg/virtualhome2kg/instance/>
PREFIX vh2kg: <http://kgrc4si.home.kg/virtualhome2kg/ontology/>
PREFIX mssn: <http://mssn.sigappfr.org/mssn/>
select DISTINCT ?action ?main_object ?target_object  where {
    ex:""" + segment + """ vh2kg:isVideoSegmentOf ?event .
    ?event vh2kg:action ?action ;
           vh2kg:mainObject ?main_object .
    OPTIONAL{?event vh2kg:targetObject ?target_object} .
}
"""

        sparql = SPARQLWrapper(ENDPOINT)
        sparql.setQuery(query)
        sparql.setReturnFormat(JSON)
        results = sparql.query().convert()

        bindings = results["results"]["bindings"]
        for result in bindings:
            action = result["action"]["value"].replace(PREFIX_VH2KG, "").replace("action/", " ")
            main_object = result["main_object"]["value"].replace(PREFIX_EX, "").replace("_" + scene, "")
            target_object = result["target_object"]["value"].replace(PREFIX_EX, "").replace("_" + scene, "") if "target_object" in result else ''
            annotation_list.append({'action': action, 'main_object': main_object, 'target_object': target_object, 'start_frame': start_frame, 'end_frame': end_frame})

    return annotation_list
